### PR TITLE
Fix for using order objects in queries

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Depends on:
- [x] #15 

In order to fix the integration tests:
1. `RepositoryTests.Should_QueryRepositoryOwner_Repositories_OrderBy_Name_Ascending`
2. `RepositoryTests.Should_QueryRepositoryOwner_Repositories_OrderBy_CreatedAt_Descending`

Which currently fail with the following exception:
```
System.AggregateException : One or more errors occurred.
---- Octokit.GraphQL.Core.GraphQLQueryException : Parse error on "." (error) at [1, 70]
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Octokit.GraphQL.IntegrationTests.RepositoryTests.Should_QueryRepositoryOwner_Repositories_OrderBy_Name_Ascending() in C:\Users\Spade\Projects\GitHub\Octokit.GraphQL\Octokit.GraphQL.IntegrationTests\RepositoryTests.cs:line 36
----- Inner Stack Trace -----
   at Octokit.GraphQL.Core.Deserializers.ResponseDeserializer.Deserialize[TResult](GraphQLQuery`1 query, JObject data)
   at Octokit.GraphQL.Core.Deserializers.ResponseDeserializer.Deserialize[TResult](GraphQLQuery`1 query, String data)
   at Octokit.GraphQL.Core.Connection.<Run>d__6`1.MoveNext()
```

From a query with the code:
```
var query = new Query().RepositoryOwner("octokit").Repositories(first: 30, orderBy: new RepositoryOrder
{
    Direction = OrderDirection.Asc,
    Field = RepositoryOrderField.Name
}).Nodes.Select(repository => repository.Name);
```

The query generated is malformed:
```
{
  repositoryOwner(login: "octokit") {
    repositories(first: 30, orderBy: Octokit.GraphQL.RepositoryOrder) {
      nodes {
        name
      }
    }
  }
}
```

The correct value of `orderBy` should be `{field: NAME, direction: ASC}`

### CODE REVIEW

**Comment by grokys** [comment]
I think the correct way to solve this is more general. `RepositoryOrder` is an `INPUT_OBJECT` type, and I suspect that the problem is we're not serializing Input Objects correctly in general. As far as I'm aware the only non-scalar types that can be passed as parameters to GraphQL fields are input objects, so we probably just need to detect that a parameter is a non-scalar type and correctly serialize its fields.

**Comment by StanleyGoldman** [comment]
Okay, I can give this another shot

File: Octokit.GraphQL.Core/Serializers/QuerySerializer.cs [hunk 1]
```diff
@@ -138,6 +138,10 @@ private string SerializeValue(object value)
             {
                 return ((bool)value) ? "true" : "false";
             }
+            else if (value.GetType().Name.EndsWith("Order"))
+            {
+                throw new NotImplementedException();
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Serializers/QuerySerializer.cs]: How would you like to identify the value as a `*Order` type?

Comment (grokys) [inline Octokit.GraphQL.Core/Serializers/QuerySerializer.cs]: It's not just `*Order` types that want to be serialized in this way - it's all [Input Objects](http://facebook.github.io/graphql/#sec-Input-Objects).

It appears that `SerializeValue` is currently only called for arguments (meaning that it should probably be renamed `SerializeArgument`) and should handle the types listed in the [spec](http://facebook.github.io/graphql/#sec-Language.Arguments): if the value being serialized isn't one of the types that map to:

`IntValue`
`FloatValue`
`StringValue`
`BooleanValue`
`NullValue`
`EnumValue`
`ListValue`

Then we can assume it's an `ObjectValue` (an Input Object) and serialize its fields.

Does that make sense?

Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Serializers/QuerySerializer.cs]: Mmm, I understand the `INPUT_OBJECT` comment you made before.
Yea that makes sense. I will throw in a bit more and bug you again.

File: Octokit.GraphQL.Core/Serializers/QuerySerializer.cs [hunk 2]
```diff
@@ -124,23 +127,57 @@ private void SerializeSelections(ISelectionSet selectionSet, StringBuilder build
...
+            else if (value is int || value is float)
+            {
+                builder.Append(value);
             }
             else
             {
-                return value.ToString();
+                var propertyInfos = value.GetType().GetRuntimeProperties()
+                    .Where(info => info.GetMethod.IsPublic);
+
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Serializers/QuerySerializer.cs]: Wanna cache this somehow?

File: Octokit.GraphQL.Core/Serializers/QuerySerializer.cs [hunk 3]
```diff
@@ -124,23 +131,64 @@ private void SerializeSelections(ISelectionSet selectionSet, StringBuilder build
             CloseBrace(builder);
         }
 
-        private string SerializeValue(object value)
+        private void SerializeValue(StringBuilder builder, object value)
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Serializers/QuerySerializer.cs]: Changed to take a `StringBuilder` as an argument so the function can be called recursively

File: Octokit.GraphQL.Core/Serializers/QuerySerializer.cs [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
+using System.Reflection;
 using System.Text;
 using Octokit.GraphQL.Core.Syntax;
+using Octokit.GraphQL.Core.Utilities;
 
 namespace Octokit.GraphQL.Core.Serializers
 {
     public class QuerySerializer
     {
+        private static readonly ConcurrentDictionary<Type, Tuple<string, MethodInfo>[]> typeCache = new ConcurrentDictionary<Type, Tuple<string, MethodInfo>[]>();
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Serializers/QuerySerializer.cs]: `ConcurrentDictionary` might be overkill here.

Comment (grokys) [inline Octokit.GraphQL.Core/Serializers/QuerySerializer.cs]: Yeah, tbh I wouldn't worry too much about caching and performance at this stage, but seeing as you've done it, looks 👍 to me.

File: Octokit.GraphQL.Core/Utilities/StringExtensions.cs [hunk 5]
```diff
@@ -1,15 +1,16 @@
 ﻿using System;
+using System.Linq;
 
 namespace Octokit.GraphQL.Core.Utilities
 {
     public static class StringExtensions
     {
-        public static string ToCamelCase(this string s)
+        public static string LowerFirstCharacter(this string s)
```
Comment (grokys) [inline Octokit.GraphQL.Core/Utilities/StringExtensions.cs]: I was using "camel case" to refer to the camel case with a lower case first letter, as opposed to "pascal case" which has the uppercase first letter, but if that seemed unclear I'm 👍 with this change.

**Comment by grokys** [comment]
LGTM! Some tests were failing due to #20 but I'm going to merge anyhow.

File: Octokit.GraphQL.Core/Utilities/StringExtensions.cs (hunk 5 above)
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Utilities/StringExtensions.cs]: As I was adding `PascalCaseToSnakeCase` I just felt over-crowded and felt the need to try to be more specific.

**Pull request merged**